tests/aws-cpp-sdk-s3-unit-tests/S3UnitTests.cpp

```diff
@@ -0,0 +1,93 @@
+
+class S3UnitTest : public testing::Test {
+protected:
+  void SetUp() override {
+    _mockClientFactory = Aws::MakeShared<MockHttpClientFactory>(ALLOCATION_TAG);
+    _mockHttpClient = Aws::MakeShared<MockHttpClient>(ALLOCATION_TAG);
+    _mockClientFactory->SetClient(_mockHttpClient);
+    HttpOptions httpOptions;
+    httpOptions.httpClientFactory_create_fn = [this]() -> std::shared_ptr<HttpClientFactory> {
+      return _mockClientFactory;
+    };
+    _options.httpOptions = httpOptions;
+    InitAPI(_options);
+    AWSCredentials credentials{"mock", "credentials"};
+    _s3Client = Aws::MakeUnique<S3Client>("ALLOCATION_TAG", credentials);
+  }
+
+  void TearDown() override {
+    ShutdownAPI(_options);
+  }
+
+  SDKOptions _options;
+  std::shared_ptr<MockHttpClient> _mockHttpClient;
+  std::shared_ptr<MockHttpClientFactory> _mockClientFactory;
+  UniquePtr<S3Client> _s3Client;
+};
+
+
+TEST_F(S3UnitTest, S3UriLeadingDots) {
+  auto putObjectRequest = PutObjectRequest()
+      .WithBucket("bluerev")
+      .WithKey("../BoredInBristol");
+
+  std::shared_ptr<IOStream> body = Aws::MakeShared<StringStream>(ALLOCATION_TAG,
+    "Bored in Bristol, always waiting",
+    std::ios_base::in | std::ios_base::binary);
+
+  putObjectRequest.SetBody(body);
+
+  //We have to mock requset because it is used to create the return body, it actually isnt used.
+  auto mockRequest = Aws::MakeShared<Standard::StandardHttpRequest>(ALLOCATION_TAG, "mockuri", HttpMethod::HTTP_GET);
+  mockRequest->SetResponseStreamFactory([]() -> IOStream* {
+    IOStream* stream = Aws::New<StringStream>(ALLOCATION_TAG, "response-string", std::ios_base::in | std::ios_base::binary);
+    return stream;
+  });
+  auto mockResponse = Aws::MakeShared<Standard::StandardHttpResponse>(ALLOCATION_TAG, mockRequest);
+  mockResponse->SetResponseCode(HttpResponseCode::OK);
+  _mockHttpClient->AddResponseToReturn(mockResponse);
+  const auto response = _s3Client->PutObject(putObjectRequest);
+  AWS_EXPECT_SUCCESS(response);
+  const auto seenRequest = _mockHttpClient->GetMostRecentHttpRequest();
+  EXPECT_EQ("https://bluerev.s3.us-east-1.amazonaws.com/../BoredInBristol", seenRequest.GetUri().GetURIString());
+}
+
+TEST_F(S3UnitTest, S3UriMiddleDots) {
+  auto putObjectRequest = PutObjectRequest()
+      .WithBucket("bluerev")
+      .WithKey("belinda/../says");
+
+  std::shared_ptr<IOStream> body = Aws::MakeShared<StringStream>(ALLOCATION_TAG,
+    "Blue Rev behind the rink I didn't really need it",
+    std::ios_base::in | std::ios_base::binary);
+
+  putObjectRequest.SetBody(body);
+
+  //We have to mock requset because it is used to create the return body, it actually isnt used.
+  auto mockRequest = Aws::MakeShared<Standard::StandardHttpRequest>(ALLOCATION_TAG, "mockuri", HttpMethod::HTTP_GET);
+  mockRequest->SetResponseStreamFactory([]() -> IOStream* {
+    return Aws::New<StringStream>(ALLOCATION_TAG, "response-string", std::ios_base::in | std::ios_base::binary);
```

Comment:
"MEMORY LEAK" -- you shout

it is deleted when `mockResponse` is delete intrun calling the destructor on its [bodyStream](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/include/aws/core/http/standard/StandardHttpResponse.h#L72), which in turn calls [delete on the underlying stream](https://github.com/aws/aws-sdk-cpp/blob/main/src/aws-cpp-sdk-core/source/utils/stream/ResponseStream.cpp#L73-L87).